src/Main.cpp

```diff
@@ -5,170 +5,126 @@
-	}
-
-	void on_tick() override
-	{
-		for (auto&& [ball_entity, ball, pos] : QueryBalls::access_storage().each())
-		{
-			for (auto&& [player_entity, pad] : QueryPlayers::access_storage().each())
-			{
-				if (intersects(pad.xy, pos.xy, 16))
-				{
-					emit(PadCollisionSignal{ ball_entity, player_entity });
-				}
-			}
-		}
-	}
-};
+	if (x > 0.0f) return  1.0f;
+	if (x < 0.0f) return -1.0f;
+	else          return  0.0f;
+}
 
-#define SCREEN_WIDTH 800
-#define SCREEN_HEIGHT 600
-#define BALL_RADIUS 16
-#define SPEED_MOD 200.0f
+static float fclampf(float x, float min, float max)
+{
+	if (x < min) return min;
+	if (x > max) return max;
+	else         return x;
+}
 
-struct BallMovementSystem
+struct MovementSystem
 	: public ecs::System
-	, public MutAccessGroupStorage<Ball, Position>
-	, public MutAccessComponentById<Ball>
-	, public SignalProcessor<PadCollisionSignal>
+	, public MutAccessGroupStorage<Movement, Position>
 {
-	virtual void process_signal(PadCollisionSignal& signal)
-	{
-		auto& ball = MutAccessComponentById<Ball>::get(signal.ball);
-		ball.speed.x *= -1;
-		ball.speed.y = get_random(-0.8f, 0.8f);
-	}
-
 	void on_tick() override
 	{
-		for (auto&& [entity, ball, pos] : access_storage().each())
+		for (auto&& [entity, movement, pos] : access_storage().each())
 		{
-			if (pos.xy.y >= SCREEN_HEIGHT - BALL_RADIUS || pos.xy.y <= BALL_RADIUS)
-			{
-				ball.speed.y *= -1;
-			}
-
-			pos.xy += ball.speed * SPEED_MOD * Time::delta_time();
+			F32 u = 16.0f;                              /* Friction coefficient. */
+			F32 m = 5.0f;                               /* Object mass (kg). */
+			F32 g = 9.807f;                             /* Gravity acceleration. */
+			F32 n = m*g;                                /* Normal force. */
+			F32 fnx = -fsignf(movement.velocity.x)*u*n; /* Friction force. */
+			F32 fny = -fsignf(movement.velocity.y)*u*n; 
+			F32 fx  = fnx+movement.force.x;             /* Total force. */
+			F32 fy  = fny+movement.force.y;
+
+			pos.xy              += movement.velocity*Time::delta_time();
+			movement.velocity.x += (fx/m)*Time::delta_time();
+			movement.velocity.y += (fy/m)*Time::delta_time();
+			movement.velocity.x = fclampf(movement.velocity.x, -movement.velocitymax.x, movement.velocitymax.x);
+			movement.velocity.y = fclampf(movement.velocity.y, -movement.velocitymax.y, movement.velocitymax.y);
 		}
 	}
 };
 
-struct PlayerControlsSystem
+struct MovementControlSystem
 	: public ecs::System
-	, public MutAccessGroupStorage<Player, KeyBindings, Position>
+	, public MutAccessGroupStorage<KeyBindings, Movement, SpriteAnimation>
 {
 	void on_tick() override
 	{
 		const auto& keys = KeyState::get();
 
-		if (keys.is_pressed(KEY_ESCAPE))
+		if (keys.is_down(KeyCode::KEY_ESCAPE))
 		{
 			Engine::get_instance().quit();
 		}
 
-		for (auto&& [entity, bindings, pos] : access_storage().each())
+		for (auto&& [entity, keybinding, movement, sprite] : access_storage().each())
```

Comment:
I'd also like `key_binding`; whenever you have multiple words, basically, cut it up